[Distributed] getNumKeys API to c10d TCPStore

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45223 [Distributed] Adding Python tests for the TCPStore getNumKeys and deleteKey
* #43963 [Distributed] DeleteKey API for c10d TCP Store
* **#43962 [Distributed] getNumKeys API to c10d TCPStore**

This PR adds a getNumKeys API to the TCP Store, which essentially returns the number of keys in the store at that point. This API will be useful for some applications related to debug logging in ProcessGroupNCCL going forward. This PR also adds some C++ tests for this API and Python tests are added in #45223. We will build on this functionality in the future by implementing this API for FileStore, HashStore, RedisStore, and ZeusStore.

Differential Revision: [D22985085](https://our.internmc.facebook.com/intern/diff/D22985085/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22985085/)!

osalpekar committed:
[Distributed] getNumKeys API to c10d TCPStore

TCPStore needs a getNumKeys API for our logging needs.

Differential Revision: [D22985085](https://our.internmc.facebook.com/intern/diff/D22985085/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22985085/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Distributed] getNumKeys API to c10d TCPStore"


TCPStore needs a getNumKeys API for our logging needs.

Differential Revision: [D22985085](https://our.internmc.facebook.com/intern/diff/D22985085/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22985085/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Distributed] getNumKeys API to c10d TCPStore"


TCPStore needs a getNumKeys API for our logging needs.

Differential Revision: [D22985085](https://our.internmc.facebook.com/intern/diff/D22985085/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22985085/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Distributed] getNumKeys API to c10d TCPStore"


TCPStore needs a getNumKeys API for our logging needs.

Differential Revision: [D22985085](https://our.internmc.facebook.com/intern/diff/D22985085/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22985085/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Distributed] getNumKeys API to c10d TCPStore"


TCPStore needs a getNumKeys API for our logging needs.

Differential Revision: [D22985085](https://our.internmc.facebook.com/intern/diff/D22985085/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22985085/)!

[ghstack-poisoned]

pritamdamania87 commented on caffe2/distributed/file_store_handler.cc:
```diff
@@ -122,6 +122,11 @@ int64_t FileStoreHandler::add(
 }
 
+int64_t FileStoreHandler::getNumKeys() {
+  CHECK(false) << "getNumKeys not implemented for FileStoreHandler";
```
shouldn't this be `TORCH_CHECK`?

osalpekar replied:
This is in the Caffe2 codebase so I don't think we should use `TORCH_CHECK`. We could use `CAFFE_ENFORCE`, but the other unimplemented function in this class used `CHECK(false)` so I used it as well.

pritamdamania87 commented on torch/lib/c10d/FileStore.cpp:
```diff
@@ -303,6 +303,10 @@ int64_t FileStore::add(const std::string& key, int64_t value) {
 }
 
+int64_t FileStore::getNumKeys() {
+  throw std::runtime_error("getNumKeys not implemented for FileStore");
```
`TORCH_CHECK` here for consistency?

pritamdamania87 commented on torch/lib/c10d/HashStore.cpp:
```diff
@@ -77,6 +77,10 @@ int64_t HashStore::add(const std::string& key, int64_t i) {
 }
 
+int64_t HashStore::getNumKeys() {
+  throw std::runtime_error("getNumKeys not implemented for HashStore");
```
`TORCH_CHECK`/

lw reviewed (commented):
Since currently this method (and deleteKey in the next pull request) is only supported in the TcpStore, does it make sense to add it to the virtual superclass? Most "users" of the stores (I am thinking of the RPC agents, but I suppose it's the same elsewhere) only accept a pointer to the virtual superclass. How can they reliably call getNumKeys on it, knowing that such a call would fail on most concrete implementations?

An alternative approach could be to add getNumKeys and deleteKey only to TcpStore, and _not_ to the superclass. If someone wants to call these methods, they can first dynamic_cast the Store pointer to a TcpStore pointer (and, by doing so, confirm that it is indeed a TcpStore and not some other type). And, if that works, they can confidently call getNumKeys.

osalpekar committed:
Update on "[Distributed] getNumKeys API to c10d TCPStore"


This PR adds a getNumKeys API to the TCP Store, which essentially returns the number of keys in the store at that point. This API will be useful for some applications related to debug logging in ProcessGroupNCCL going forward. This PR also adds some C++ tests for this API and Python tests are added in #45223. We will build on this functionality in the future by implementing this API for FileStore, HashStore, RedisStore, and ZeusStore.

Differential Revision: [D22985085](https://our.internmc.facebook.com/intern/diff/D22985085/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22985085/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Distributed] getNumKeys API to c10d TCPStore"


This PR adds a getNumKeys API to the TCP Store, which essentially returns the number of keys in the store at that point. This API will be useful for some applications related to debug logging in ProcessGroupNCCL going forward. This PR also adds some C++ tests for this API and Python tests are added in #45223. We will build on this functionality in the future by implementing this API for FileStore, HashStore, RedisStore, and ZeusStore.

Differential Revision: [D22985085](https://our.internmc.facebook.com/intern/diff/D22985085/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22985085/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Distributed] getNumKeys API to c10d TCPStore"


This PR adds a getNumKeys API to the TCP Store, which essentially returns the number of keys in the store at that point. This API will be useful for some applications related to debug logging in ProcessGroupNCCL going forward. This PR also adds some C++ tests for this API and Python tests are added in #45223. We will build on this functionality in the future by implementing this API for FileStore, HashStore, RedisStore, and ZeusStore.

Differential Revision: [D22985085](https://our.internmc.facebook.com/intern/diff/D22985085/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22985085/)!

[ghstack-poisoned]
